src/System.Composition.Hosting/src/System/Composition/Hosting/Providers/ExportFactory/ExportFactoryExportDescriptorProvider.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Linq;
+using System.Reflection;
+using System.Composition.Hosting.Core;
+using System.Composition.Runtime;
+using System.Collections.Generic;
+using System.Composition.Hosting.Util;
+using System.Composition.Hosting.Properties;
+
+namespace System.Composition.Hosting.Providers.ExportFactory
+{
+    internal class ExportFactoryExportDescriptorProvider : ExportDescriptorProvider
+    {
+        private static readonly MethodInfo s_getExportFactoryDefinitionsMethod = typeof(ExportFactoryExportDescriptorProvider).GetTypeInfo().GetDeclaredMethod("GetExportFactoryDescriptors");
+
+        public override IEnumerable<ExportDescriptorPromise> GetExportDescriptors(CompositionContract exportKey, DependencyAccessor definitionAccessor)
+        {
+            if (!exportKey.ContractType.IsConstructedGenericType || exportKey.ContractType.GetGenericTypeDefinition() != typeof(ExportFactory<>))
+                return NoExportDescriptors;
+
+            var gld = s_getExportFactoryDefinitionsMethod.MakeGenericMethod(exportKey.ContractType.GenericTypeArguments[0]);
+            var gldm = gld.CreateStaticDelegate<Func<CompositionContract, DependencyAccessor, object>>();
+            return (ExportDescriptorPromise[])gldm(exportKey, definitionAccessor);
+        }
+
+        private static ExportDescriptorPromise[] GetExportFactoryDescriptors<TProduct>(CompositionContract exportFactoryContract, DependencyAccessor definitionAccessor)
+        {
+            var productContract = exportFactoryContract.ChangeType(typeof(TProduct));
+            var boundaries = new string[0];
```

Comment:
@stephentoub I've added `EmptyArray<T>` to the Common folder, and I think I've updated all instances of creating an empty array to use it.